File: C++/Backtracking/491. Non-decreasing Subsequences .cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<vector<int>> res;

    vector<vector<int>> findSubsequences(vector<int> &nums) {
        vector<int> cur;
        backtracking(cur, nums, 0);
        return res;
    }

    void backtracking(vector<int> &cur, vector<int> &nums, int idx) {
        if (cur.size() > 1) res.push_back(cur);
        unordered_set<int> seen;
        for (int i = idx; i < nums.size(); ++i) {
            if ((cur.empty() || cur.back() <= nums[i]) && !seen.count(nums[i])) {
                cur.push_back(nums[i]);
                backtracking(cur, nums, i + 1);
                cur.pop_back();
                seen.insert(nums[i]);
            }
        }
    }
};

File: Python/backtracking/491. Non-decreasing Subsequences.py
from typing import List


class Solution:
    def findSubsequences(self, nums: List[int]) -> List[List[int]]:
        res = set()

        def backtracking(cur, idx):
            nonlocal res
            if len(cur) > 1:
                res.add(tuple(cur))
            if idx >= len(nums): return

            if not cur or nums[idx] >= cur[-1]:
                backtracking(cur + [nums[idx]], idx + 1)
            backtracking(cur, idx + 1)

        backtracking([], 0)
        return res
